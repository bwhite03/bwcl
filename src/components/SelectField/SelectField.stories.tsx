import React from "react";
import { Story, Meta } from "@storybook/react";
import SelectField, { SelectFieldProps } from "./SelectField";

const fakeData = [
  {
    id: 1,
    name: "tim",
    age: 25,
  },
  {
    id: 2,
    name: "sally",
    age: 23,
  },
];

export default {
  title: "Forms/Select",
  component: SelectField,
} as Meta;

const Template =
  <T extends {}>(): Story<SelectFieldProps<T>> =>
  (args) =>
    <SelectField<T> {...args} />;

export const SelectFieldExample = Template<object>().bind({});
SelectFieldExample.args = {
  id: "s1",
  label: "select a  State",
  name: "s1",
  data: fakeData,
  emptyMsg: "select a state",
  // @ts-ignore
  displayField: "name",
  value: "id",
};
