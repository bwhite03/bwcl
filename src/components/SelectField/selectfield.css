.bwcl-select select {
  padding: 4px 8px;
  border-radius: 8px;
  outline: none;
}

.bwcl-select,
.alert-danger {
  background-color: red;
  color: white;
  border-radius: 5px;
}
