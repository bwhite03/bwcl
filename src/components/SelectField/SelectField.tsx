import React from "react";
import "./selectfield.css";

export interface SelectFieldProps<T> {
  id: string;
  name: string;
  label: string;
  onChange: (e: React.ChangeEvent<HTMLSelectElement>) => string;
  valueField: T[keyof T];
  displayField: keyof T;
  error?: string;
  emptyMsg: string;
  data: T[];
}

function SelectField<T>(props: SelectFieldProps<T>) {
  let wrapperClass = "bwcl-select";
  if (props.error && props.error.length > 0) {
    wrapperClass += " has-error";
  }

  return (
    <div className={wrapperClass}>
      <label htmlFor={props.id}>{props.label}</label>
      <div className="field">
        <select
          id={props.id}
          onChange={props.onChange}
          className="form-control"
        >
          <option value="0">{props.emptyMsg}</option>
          {props.data && props.data.length === 0
            ? null
            : props.data.map((item: T, index: number) => (
                <option
                  key={`${props.id}-${index}`}
                  // @ts-ignore
                  value={item[props.valueField]}
                >
                  {item[props.displayField]}
                </option>
              ))}
        </select>
      </div>
      {props.error && (
        <div className="alert-danger" data-testid="alert">
          {props.error}
        </div>
      )}
    </div>
  );
}

export default SelectField;
