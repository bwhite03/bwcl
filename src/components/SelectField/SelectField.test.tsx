import React from "react";
import { render } from "@testing-library/react";
import SelectField from "./SelectField";

describe("SelectField", () => {
  test("component renders correctly", () => {
    const testFn = jest.fn();
    const fakeData = [
      {
        id: 1,
        name: "tim",
        age: 25,
      },
      {
        id: 2,
        name: "sally",
        age: 23,
      },
    ];
    const frag = render(
      <SelectField
        id="test"
        label="test"
        name="test"
        onChange={testFn}
        data={fakeData}
        displayField="name"
        valueField="id"
        emptyMsg="select a person"
      />
    );
    expect(frag).toMatchSnapshot();
  });
});
