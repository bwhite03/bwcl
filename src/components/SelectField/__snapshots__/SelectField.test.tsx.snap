// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField component renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="form-group"
      >
        <label
          for="test"
        >
          test
        </label>
        <div
          class="field"
        >
          <select
            class="form-control"
            id="test"
          >
            <option
              value="0"
            >
              select a person
            </option>
            <option
              value="1"
            >
              tim
            </option>
            <option
              value="2"
            >
              sally
            </option>
          </select>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="form-group"
    >
      <label
        for="test"
      >
        test
      </label>
      <div
        class="field"
      >
        <select
          class="form-control"
          id="test"
        >
          <option
            value="0"
          >
            select a person
          </option>
          <option
            value="1"
          >
            tim
          </option>
          <option
            value="2"
          >
            sally
          </option>
        </select>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SelectField should render with an empty error 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="form-group"
      >
        <label
          for="test"
        >
          test
        </label>
        <div
          class="field"
        >
          <select
            class="form-control"
            id="test"
          >
            <option
              value="0"
            >
              select a person
            </option>
            <option
              value="1"
            >
              tim
            </option>
            <option
              value="2"
            >
              sally
            </option>
          </select>
        </div>
        
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="form-group"
    >
      <label
        for="test"
      >
        test
      </label>
      <div
        class="field"
      >
        <select
          class="form-control"
          id="test"
        >
          <option
            value="0"
          >
            select a person
          </option>
          <option
            value="1"
          >
            tim
          </option>
          <option
            value="2"
          >
            sally
          </option>
        </select>
      </div>
      
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SelectField should render with an error 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="form-group has-error"
      >
        <label
          for="test"
        >
          test
        </label>
        <div
          class="field"
        >
          <select
            class="form-control"
            id="test"
          >
            <option
              value="0"
            >
              select a person
            </option>
            <option
              value="1"
            >
              tim
            </option>
            <option
              value="2"
            >
              sally
            </option>
          </select>
        </div>
        <div
          class="alert alert-danger"
        >
          error
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="form-group has-error"
    >
      <label
        for="test"
      >
        test
      </label>
      <div
        class="field"
      >
        <select
          class="form-control"
          id="test"
        >
          <option
            value="0"
          >
            select a person
          </option>
          <option
            value="1"
          >
            tim
          </option>
          <option
            value="2"
          >
            sally
          </option>
        </select>
      </div>
      <div
        class="alert alert-danger"
      >
        error
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
